File: 46. Permutations/Solution.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        if (nums.size() == 0) {
            vector<int> r;
            res.push_back(r);
            return res;
        }
        vector<int> foll(nums.size() - 1, 0);
        int push = 0;
        for (int i = 0; i < nums.size(); i++) {
            for (int j = 0; j < nums.size(); j++) {
                if (i == j) {
                    push--;
                    continue;
                }
                foll[j + push] = nums[j];
            }
            push = 0;
            for (vector<int> prev : permute(foll)) {
                prev.push_back(nums[i]);
                res.push_back(prev);
            }
        }
        return res;
    }
};

File: 46. Permutations/Solution.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        if (nums.length == 0) {
            res.add(new ArrayList<>());
            return res;
        }
        int[] foll = new int[nums.length - 1];
        int push = 0;
        for (int i = 0; i < nums.length; i++) {
            for (int j = 0; j < nums.length; j++) {
                if (i == j) {
                    push--;
                    continue;
                }
                foll[j + push] = nums[j];
            }
            push = 0;
            for (List<Integer> prev : permute(foll)) {
                prev.add(nums[i]);
                res.add(prev);
            }
        }
        return res;
    }
}
